examples/jsm/loaders/UltraHDRLoader.js

```diff
@@ -0,0 +1,592 @@
+
+				}
+
+				const canvas = document.createElement( 'canvas' );
+				const ctx = canvas.getContext( '2d', {
+					willReadFrequently: true,
+					colorSpace: 'srgb',
+				} );
+
+				canvas.width = sdrImage.width;
+				canvas.height = sdrImage.height;
+
+				/* Use out-of-the-box interpolation of Canvas API to scale gainmap to fit the SDR resolution */
+				ctx.drawImage(
+					gainmapImage.source,
+					0,
+					0,
+					gainmapImage.width,
+					gainmapImage.height,
+					0,
+					0,
+					sdrImage.width,
+					sdrImage.height
+				);
+				const gainmapImageData = ctx.getImageData(
+					0,
+					0,
+					sdrImage.width,
+					sdrImage.height,
+					{ colorSpace: 'srgb' }
+				);
+
+				ctx.drawImage( sdrImage.source, 0, 0 );
+				const sdrImageData = ctx.getImageData(
+					0,
+					0,
+					sdrImage.width,
+					sdrImage.height,
+					{ colorSpace: 'srgb' }
+				);
+
+				/* HDR Recovery formula - https://developer.android.com/media/platform/hdr-image-format#use_the_gain_map_to_create_adapted_HDR_rendition */
+				let hdrBuffer;
+
+				if ( this.type === HalfFloatType ) {
+
+					hdrBuffer = new Uint16Array( sdrImageData.data.length ).fill( 23544 );
+
+				} else {
+
+					hdrBuffer = new Float32Array( sdrImageData.data.length ).fill( 255 );
+
+				}
+
+				const maxDisplayBoost = Math.sqrt(
+					Math.pow(
+						/* 1.8 instead of 2 near-perfectly rectifies approximations introduced by precalculated SRGB_TO_LINEAR values */
+						1.8,
+						xmpMetadata.hdrCapacityMax
+					)
+				);
+				const unclampedWeightFactor =
+					( Math.log2( maxDisplayBoost ) - xmpMetadata.hdrCapacityMin ) /
+					( xmpMetadata.hdrCapacityMax - xmpMetadata.hdrCapacityMin );
+				const weightFactor = Math.min(
+					Math.max( unclampedWeightFactor, 0.0 ),
+					1.0
+				);
+				const useGammaOne = xmpMetadata.gamma === 1.0;
+
+				for (
+					let pixelIndex = 0;
+					pixelIndex < sdrImageData.data.length;
+					pixelIndex += 4
+				) {
+
+					const x = ( pixelIndex / 4 ) % sdrImage.width;
+					const y = Math.floor( pixelIndex / 4 / sdrImage.width );
+
+					for ( let index = pixelIndex; index < pixelIndex + 3; index ++ ) {
+
+						const sdrValue = sdrImageData.data[ index ];
+
+						const gainmapIndex = ( y * sdrImage.width + x ) * 4;
+						const gainmapValue = gainmapImageData.data[ gainmapIndex ] / 255.0;
+
+						/* Gamma is 1.0 by default */
+						const logRecovery = useGammaOne
+							? gainmapValue
+							: Math.pow( gainmapValue, 1.0 / xmpMetadata.gamma );
+
+						const logBoost =
+							xmpMetadata.gainMapMin * ( 1.0 - logRecovery ) +
+							xmpMetadata.gainMapMax * logRecovery;
+
+						const hdrValue =
+						( sdrValue + xmpMetadata.offsetSDR ) *
+						( logBoost * weightFactor === 0.0
+							? 1.0
+							: Math.pow( 2, logBoost * weightFactor ) ) -
+					xmpMetadata.offsetHDR;
+
+						const linearHDRValue = Math.min(
+							Math.max(
+								this._srgbToLinear( hdrValue ),
+								0
+							),
+							65504
+						);
```

Comment:
The color grading issue was with the channel index not being applied to the gainmap image (gainmap loop used only the red channel because of that 🙇‍♂️ )

As for SRGB-to-Linear - according to spec, that's indeed correct. Only the SDR image should be converted and gainmap image is assumed to be linear right away. The reason for moving it to the last step is in Canvas API ([here](https://github.com/mrdoob/three.js/pull/28825/files#diff-5022d5b4ab0557dae9c84141861186868879339d91728f4c7c863f4ef3af7c25R482)) - currently it allows only rendering colors in [SRGB & Display-P3](https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/getContext#colorspace), which I think accidentally converts the gainmap from LinearSRGB to SRGB (either this step or the loading the image via `<img>`) 👀

Reply:
Note that the gainmap is neither sRGB nor Linear-sRGB, it is non-color data. When the gainmap is drawn to the Canvas API, I'd expect the Canvas API to either (a) observe no color space (as it's non-color data) and draw the gainmap without conversion, or (b) assume the gainmap is already sRGB and (again) draw the gainmap without conversion. In either case, calling `getImageData` with `colorSpace: 'srgb'` is the right choice, as it again avoids any conversions, which would inherently be unwanted on non-color data.

Which is not to say that something in the loading process couldn't be applying an unwanted conversion ... the web platform does not really understand either Linear-sRGB or non-color identifiers well ... but I think we can avoid that if so.